### ISSUE TITLE
Support opening relative links

### ISSUE THREAD

**Issue by amake** [amake/orgro#5]
It would be nice to support opening relative file links like `[[./other_file.org][related text]]`.

Issues:

- Is it even possible to resolve a URI relative to one that we have obtained permission for?
  - Maybe we can prompt for permissions, e.g. show the file picker with the target pre-selected?
- How should the file open?
  - Push a new route on top of the current?
  - Or close the current and open the new file?

**Comment by alex-eyre** [comment]
I know apps like Syncthing use the new scoped-storage API to request access to a whole folder, so that's an option for addressing issue 1.
imho the file should close the current file and open the new one in it's place, rather than trying to implement some kind of slide-over system, but that's just my two-cents.

**Comment by fgilbert68** [comment]
I would really like this. I heavily rely on links like `[[other_file.org::*Some outline][related text]]` to switch between ideas and the Org file where they are written.
For me, the natural way would be to close the current file and open the new one; displaying a "narrow" view if the link includes an outline.

I'm not sure I understand the URI issue (but I don't know anything about the apparently very complex world of Android permissions). All my org files are in the same directory on the SD card (copied there using rsync through a davfs mount...), and if this is the case (as you suggested by writing `[[./other_file.org]]`), no extra permission should be needed?

**Comment by amake** [comment]
> All my org files are in the same directory on the SD card (copied there using rsync through a davfs mount...), and if this is the case (as you suggested by writing [[./other_file.org]]), no extra permission should be needed?

Right now Orgro can only obtain permissions to individual files. So to access a separate file I would need to make you choose it from a picker. If I can preselect the file that might not be so bad. If I cache the access ID then you'd only have to choose it once.

But then you're still choosing each new file.

I'm not sure if obtaining permission to a directory would let me access all of its children. It might. But then there is a new concept of "choosing a directory" to somehow include in the UI and make it clear how/why you need to do that. That might require redoing all of the file-opening workflow code.

So this ticket is stuck waiting for me to have time to investigate all of this.

**Comment by mcubrilo** [comment]
bump

**Comment by MisterY** [comment]
Very briefly, I can confirm that Flutter on Android can obtain permissions for a folder on external storage. Just recently I worked on the related code in GitJournal. This allows opening any file within that the hierarchy below the selected location.

I'd agree that a more-encompassing strategy regarding file-opening would be worthwhile. I'm using Orgro as the viewer for the files edited by Orgzly, located in a local repository. Like others, the folder tree is synchronized via rclone.

Perhaps an option to "add a folder" could simply display all the .org files within?
The issue with the current file-based permission from the content provider is that the permission is void as soon as the file/content is edited. After each edit, I need to re-select the file to preview it. Hopefully I still remember where exactly it came from.
Going further, iterating through a folder tree and displaying notes (hopefully in a tree) would solve that issue for me.

**Comment by amake** [comment]
For #22 I've hit upon what I consider a plausible UI for requesting document-specific access permissions, so this is partially unblocked.

My idea is to detect at open time whether there are relative links, and to compute the common root directory.
- If we know we have read access to that directory, then we do nothing as the links are already resolvable.
- If we don't have read access, we show a banner at the top of the document notifying the user:

    ```
    This document contains relative links. Allow access?
    [Dismiss] [Allow]
    ```

    Tap `Allow` to show a directory picker (hopefully) already navigated to the correct directory. If the user selects the directory, store the persistent bookmark for future use so that the banner doesn't need to be shown again.

Remaining issues:

# Getting permissions

I use [file_picker_writable](https://pub.dev/packages/file_picker_writable) to get persistent access to files, but this plugin doesn't allow selecting directories. So I either need to add the functionality to that plugin, or write my own (see https://github.com/hpoul/file_picker_writable/issues/16).

# Handling permissions

Directory permissions should be persisted so that the user doesn't get asked every time.

Given a set of existing permissions and a set of URLs, it might be non-trivial to compute whether we need to ask for additional permissions.

Permissions may expire or otherwise become temporarily or permanently unusable (switching users in Google Drive); figuring out how to cull them might be important.

# Error handling

- How to handle links that aren't resolvable
- How to handle persisted permissions that have become stale
- How to handle scenarios where the base URL is not known (opening via intent)

**Comment by MisterY** [comment]
For getting permissions, I would point you to the code in GitJournal when the user changes the location of the repository. I could find the exact file, if needed. I think it is in the Settings form.  This definitely works on Android, at least.

**Comment by MisterY** [comment]
> I don't want to change the topmost file-picking UI to introduce the concept of directories, or implement some sort of generalized directory tree browser.

Interesting, since you mentioned you want to stick to the original OrgMode features yet here you reject the most basic one? OrgMode is defacto file-based, therefore directory-based, system of notes. Why the disconnect?

**Comment by amake** [comment]
> OrgMode is defacto a file-based, therefore directory-based, system of notes. Why the disconnect?

I disagree with this premise. Org Mode is *buffer*-based. Buffers are often backed by files, but don't have to be. Links (relative and otherwise) *can* point to files, but don't have to. You could have Org Mode function exactly as-is but have all the data come out of a database, or be fetched over the network, or really anything, as long as the links resolve by *some* mechanism.

There is nothing fundamental to Org Mode about either files or directories.

> I'm asking because it seems that most of the open issues would be resolved by implementing a folder-based structure.

- The OS already has a file/directory browser that the user already has to interact with; mine would be reinventing the wheel (probably poorly)
- Making an in-app directory browser doesn't solve the issue.

    Even if I make the user choose a directory and then a file, thus ensuring access to the filesystem tree containing the file, the file may reference other files outside of that tree (e.g. `../foo.org`). We may need to start arbitrarily far up the tree; the only way to find out is to open the file and inspect the links within.

**Comment by amake** [comment]
I have this implemented locally to some degree, but there are large limitations.

# Android-specific

It appears to be a fundamental limitation of the Android Storage Access Framework that you can't resolve *up* the tree from a known content URI.

If you open a file `/a/b/foo.org` that has a relative link to `../c/bar.org`, the app must prompt for access permissions to `/a`, but given the content URI for `foo.org` (obtained from a file picker) and the relative path to `bar.org` (obtained from the content of `foo.org`), there seems to be no way to get a content URI for `/a` or even discover the name `a`. This means that prompting for access has very poor UX: I can't pre-select `/a` for the user or even tell them to select `/a`; I can only say "select the directory corresponding to `..` from the file `foo.org`.

Either that, or I have to ask for permission to manage the entire provider document tree.

# iOS and Android

Very few content providers implement support for picking a directory (`ACTION_OPEN_DOCUMENT_TREE` on Android; `kUTTypeFolder` on iOS). The only ones I've managed to find so far are:

- iOS
  - Local storage
  - iCloud Storage
- Android
  - Local storage
  - Termux
  - Seafile

(Providers that definitely *don't* support it include Google Drive and Dropbox.)

So opening relative links will be limited to supported providers, but I don't know of a way to programmatically determine if a provider supports it, so there will be some annoyances.

**Comment by amake** [comment]
This will be implemented in v1.18.0, which will be available for testing soon:
- [Android](https://play.google.com/apps/testing/com.madlonkay.orgro)
- [iOS](https://testflight.apple.com/join/8YTxcuZC)

See [here](https://orgro.org/faq/#how-can-i-use-relative-links-between-org-mode-files) for details and caveats.

**Comment by amake** [comment]
v1.18.0 is now available for testing on Test Flight and Google Play.

**Comment by amake** [comment]
This is implemented in v1.18.1, which is out for all platforms.
